mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp

```diff
@@ -742,6 +747,197 @@ extractLayouts(Operation *op, llvm::StringMap<unsigned> &fLayoutMap,
   return success();
 }
 
+// Structure to hold information about a single reduction operation
+struct ReductionInfo {
+  ReduceMethod method;
+  int64_t axis;
+  int64_t rank;
+  int64_t stride; // Stride of the reduction dimension
+  bool hasPointwiseBefore;
+
+  bool operator<(const ReductionInfo &other) const {
+    // Sort by method first, then rank, then axis, then stride, then
+    // hasPointwiseBefore
+    if (method != other.method)
+      return method < other.method;
+    if (rank != other.rank)
+      return rank < other.rank;
+    if (axis != other.axis)
+      return axis < other.axis;
+    if (stride != other.stride)
+      return stride < other.stride;
+    return hasPointwiseBefore > other.hasPointwiseBefore;
+  }
+
+  bool operator==(const ReductionInfo &other) const {
+    return method == other.method && axis == other.axis && rank == other.rank &&
+           stride == other.stride &&
+           hasPointwiseBefore == other.hasPointwiseBefore;
+  }
+};
+
+// Structure to hold fusion information for problem key generation
+struct FusionInfo {
+  SmallVector<ReductionInfo> reductions;
+
+  bool hasReduction() const { return !reductions.empty(); }
+  int numReductionOutputs() const { return reductions.size(); }
```

Comment:
nit:
```suggestion
  int numReductions() const { return reductions.size(); }
```

numReductionOutputs is ambigous. It could mean how many reduction operations are being returned from the function or how many outputs each reduction op has (if there is such an op with multiple outputs).